llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,567 @@
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode);
+      size_t End = CurrentPtr;
+      SmallString<0> RawBody;
+      for (std::size_t I = Start; I < End - 1; I++)
+        RawBody += Tokens[I].getRawBody();
+      CurrentNode->setRawBody(RawBody);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::InvertSectionOpen: {
+      CurrentNode = new (Node) ASTNode(ASTNode::InvertSection, A, Parent);
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode);
+      size_t End = CurrentPtr;
+      SmallString<0> RawBody;
+      for (size_t Idx = Start; Idx < End - 1; Idx++)
+        RawBody += Tokens[Idx].getRawBody();
+      CurrentNode->setRawBody(RawBody);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::Comment:
+      break;
+    case Token::Type::SectionClose:
+      return;
+    }
+  }
+}
+
+StringRef Template::render(Value &Data) {
+  Tree->render(Data, Output);
+  LocalAllocator.Reset();
+  return Output.str();
+}
+
+void Template::registerPartial(StringRef Name, StringRef Partial) {
+  Parser P = Parser(Partial, Allocator);
+  ASTNode *PartialTree = P.parse();
+  PartialTree->setUpNode(LocalAllocator, Partials, Lambdas, SectionLambdas,
+                         Escapes);
+  Partials.insert(std::make_pair(Name, PartialTree));
+}
+
+void Template::registerLambda(StringRef Name, Lambda L) { Lambdas[Name] = L; }
+
+void Template::registerLambda(StringRef Name, SectionLambda L) {
+  SectionLambdas[Name] = L;
+}
+
+void Template::registerEscape(DenseMap<char, StringRef> E) { Escapes = E; }
+
+Template::Template(StringRef TemplateStr) {
+  Parser P = Parser(TemplateStr, Allocator);
+  Tree = P.parse();
+  // the default behaviour is to escape html entities
+  DenseMap<char, StringRef> HtmlEntities = {{'&', "&amp;"},
+                                            {'<', "&lt;"},
+                                            {'>', "&gt;"},
+                                            {'"', "&quot;"},
+                                            {'\'', "&#39;"}};
+  registerEscape(HtmlEntities);
+  Tree->setUpNode(LocalAllocator, Partials, Lambdas, SectionLambdas, Escapes);
+}
+
+void toJsonString(const Value &Data, SmallString<0> &Output) {
+  if (Data.getAsNull())
+    return;
+  if (auto *Arr = Data.getAsArray())
+    if (Arr->empty())
+      return;
+  if (Data.getAsString()) {
+    Output = Data.getAsString()->str();
+    return;
+  }
+  if (auto Num = Data.getAsNumber()) {
+    std::ostringstream Oss;
+    Oss << *Num;
+    Output = Oss.str();
+    return;
+  }
+  Output = formatv("{0:2}", Data);
+}
+
+bool isFalsey(const Value &V) {
+  return V.getAsNull() || (V.getAsBoolean() && !V.getAsBoolean().value()) ||
+         (V.getAsArray() && V.getAsArray()->empty()) ||
+         (V.getAsObject() && V.getAsObject()->empty());
+}
+
+void ASTNode::render(const Value &Data, SmallString<0> &Output) {
```

Comment:
Fro this and the various `render*` APIs is there much/any value in having an in/out parameter? most of the cases here overwrite Output, not append to it, and I wonder if it would be easier/nicer to just return an `std::string`. For functions that want to append its possible to use a different API.

Alternatively, using one of the OStreams like @cor3ntin mentioned in more places could be a good alternative to passing the output buffer around.

Reply:
I think you're viewing an outdated version of the PR I've already refactor this function to use OStream?

Reply:
hmm, seems so. I only opened this PR this morning though. Since that's done, feel free to disregard.